File: Dynamic Programming/maximum_subarray.py
class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        dp = nums
        max_sum = dp[0]
      
        for i in range(1, len(nums)):
            dp[i] = max(dp[i], dp[i - 1] + dp[i])

            if dp[i] > max_sum:
                max_sum = dp[i]

        return max_sum

File: Array & Strings/maximum_subarray.cpp
class Solution {
public:
    int maxSubArray(vector<int>& nums) {
        int current_index_sum = nums[0], overall_max_sum = nums[0];

        // Kadane's algorithm
        for(int i = 1; i < nums.size(); i++) {
            current_index_sum = max(nums[i], current_index_sum + nums[i]);
            overall_max_sum = max(overall_max_sum, current_index_sum);
        }
        return overall_max_sum;
    }
};
